// 8base-js-sdk/lib/graphql.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.USER_LOGIN_MUTATION = `
  mutation UserLoginMutation($data: UserLoginInput!) {
    userLogin(data: $data) {
      success
      auth {
        idToken
        refreshToken
      }
    }
  }
`;
exports.USER_SIGN_UP_MUTATION = `
  mutation UserSignUpMutation($authProfileId: ID!, $password: String!, $user: UserCreateInput!) {
    userSignUpWithPassword(
      authProfileId: $authProfileId,
      password: $password,
      user: $user
    ) {
      id
      email
      firstName
      lastName
      timezone
      status
    }
  }
`;
exports.USER_SIGN_UP_WITH_TOKEN_MUTATION = `
  mutation UserSignUpWithTokenMutation($authProfileId: ID!, $user: UserCreateInput!) {
    userSignUpWithToken(
      authProfileId: $authProfileId,
      user: $user
    ) {
      id
      email
      firstName
      lastName
      timezone
      status
    }
  }
`;
exports.CURRENT_USER_QUERY = `
  query {
    user {
      id
      email
      firstName
      lastName
      timezone
      avatar {
        downloadUrl
      }
    }
  }
`;
