// @8base-js-sdk/api/lib/ErrorHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const constants_1 = require("./constants");
class ErrorHandler {
    constructor(callback) {
        if (typeof callback === 'function') {
            this.errorCallback = callback;
        }
        else if (callback) {
            this.errorMap = callback;
        }
    }
    handle(error, rerun) {
        if (this.errorCallback) {
            return this.handleWithCallback(error, rerun);
        }
        if (this.errorMap) {
            return this.handleWithErrorMap(error, rerun);
        }
    }
    handleWithCallback(error, rerun) {
        if (!this.errorCallback) {
            return;
        }
        return this.errorCallback(error, rerun);
    }
    handleWithErrorMap(error, rerun) {
        if (!this.errorMap) {
            return;
        }
        const code = error.response.errors[0].code;
        if (code && typeof this.errorMap[code] === 'function') {
            return this.errorMap[code](error, rerun);
        }
        else if (typeof this.errorMap[constants_1.DEFAULT_ERROR_MAP_KEY] === 'function') {
            return this.errorMap[constants_1.DEFAULT_ERROR_MAP_KEY](error, rerun);
        }
    }
}
exports.ErrorHandler = ErrorHandler;
